src/main/scala/encry/view/state/avlTree/AvlTree.scala

```diff
@@ -0,0 +1,411 @@
+  }
+
+  private def getLeftPath(node: Node[K, V]): List[Node[K, V]] = node match {
+    case shadowNode: ShadowNode[K, V] =>
+      val restoredNode = shadowNode.restoreFullNode(storage)
+      getLeftPath(restoredNode)
+    case leafNode: LeafNode[K, V] => List(leafNode)
+    case internalNode: InternalNode[K, V] => internalNode +: internalNode.leftChild.map(getLeftPath).getOrElse(List.empty)
+  }
+
+  private def insert(newKey: K,
+                     newValue: V,
+                     nodeWithOpInfo: NodeWithOpInfo[K, V]): NodeWithOpInfo[K, V] = nodeWithOpInfo.node match {
+    case shadowNode: ShadowNode[K, V] =>
+      val restoredNode = shadowNode.restoreFullNode(storage)
+      insert(newKey, newValue, NodeWithOpInfo(restoredNode, nodeWithOpInfo.opInfo))
+    case _: EmptyNode[K, V] =>
+      val newLeaf = LeafNode[K, V](newKey, newValue)
+      NodeWithOpInfo(newLeaf, nodeWithOpInfo.opInfo.updateInserted(ByteArrayWrapper(newLeaf.hash) -> newLeaf))
+    case leafNode: LeafNode[K, V] =>
+      if (leafNode.key === newKey) NodeWithOpInfo(leafNode.copy(value = newValue), nodeWithOpInfo.opInfo)
+      else {
+        val newInternalNode = InternalNode[K, V](leafNode.key, leafNode.value, height = 1, balance = 0)
+        insert(
+          newKey,
+          newValue,
+          NodeWithOpInfo(
+            newInternalNode,
+            nodeWithOpInfo.opInfo.update(ByteArrayWrapper(newInternalNode.hash) -> newInternalNode, ByteArrayWrapper(leafNode.hash))
+          )
+        )
+      }
+    case internalNode: InternalNode[K, V] =>
+      if (internalNode.key > newKey) {
+        val newLeftChild = internalNode.leftChild.map(previousLeftChild =>
+          insert(newKey, newValue, nodeWithOpInfo.copy(node = previousLeftChild))
+        ).getOrElse{
+          val newLeaf = LeafNode(newKey, newValue)
+          NodeWithOpInfo(newLeaf, nodeWithOpInfo.opInfo.updateInserted(ByteArrayWrapper(newLeaf.hash) -> newLeaf))
+        }
+        val newNode = internalNode.updateChilds(newLeftChild = Some(newLeftChild.node), prevOpsInfo = newLeftChild.opInfo)
+        balance(newNode.copy(opInfo = newNode.opInfo.update(ByteArrayWrapper(newNode.node.hash) -> newNode.node, ByteArrayWrapper(internalNode.hash))))
+      } else {
+        val newRightChild = internalNode.rightChild.map(previousRightChild =>
+          insert(newKey, newValue, nodeWithOpInfo.copy(node = previousRightChild))
+        ).getOrElse{
+          val newLeaf = LeafNode(newKey, newValue)
+          NodeWithOpInfo(newLeaf, nodeWithOpInfo.opInfo.updateInserted(ByteArrayWrapper(newLeaf.hash) -> newLeaf))
+        }
+        val newNode =
+          internalNode.updateChilds(newRightChild = Some(newRightChild.node), prevOpsInfo = newRightChild.opInfo)
+        balance(newNode.copy(opInfo = newNode.opInfo.update(ByteArrayWrapper(newNode.node.hash) -> newNode.node, ByteArrayWrapper(internalNode.hash))))
```

Comment:
You are calling `balance` here with exact same arguments as on line 244, so, you can move it out of if statement and then call